File: day11/day11.cpp
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

static const pair<int, int> NEIGHBORS[] = {
	{-1, 0},
	{-1, -1},
	{0, -1},
	{1, -1},
	{1, 0},
	{1, 1},
	{0, 1},
	{-1, 1},
};

struct Map
{
	char data[10][10];

	void inc(int x, int y)
	{
		if (data[y][x] > 0 && ++data[y][x] > '9')
		{
			data[y][x] = 0;
			for (auto [dx, dy]: NEIGHBORS)
			{
				int nx = x + dx;
				int ny = y + dy;
				if (0 <= nx && nx < 10 && 0 <= ny && ny < 10)
				{
					inc(nx, ny);
				}
			}
		}
	}

	size_t step()
	{
		for (int y = 0; y < 10; y++)
		{
			for (int x = 0; x < 10; x++)
			{
				inc(x, y);
			}
		}

		size_t count = 0;
		for (int y = 0; y < 10; y++)
		{
			for (int x = 0; x < 10; x++)
			{
				if (data[y][x] == 0)
				{
					data[y][x] = '0';
					count++;
				}
			}
		}
		return count;
	}
};

istream& operator>>(istream& in, Map &m)
{
	int y = 0;
	int x = 0;
	char ch;
	while (y < 10 && in.get(ch))
	{
		if (ch == '\n')
		{
			y++;
			x = 0;
		}
		else if (x < 10)
		{
			m.data[y][x] = ch;
			x++;
		}
	}
	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}
	Map m;
	in >> m;
	in.close();

	size_t part1 = 0;
	size_t part2 = 0;
	while (1)
	{
		size_t flashes = m.step();
		if (part2 < 100)
		{
			part1 += flashes;
		}
		part2++;
		if (flashes == 100)
		{
			break;
		}
	}

	fmt::print("Part1: {}\n", part1);
	fmt::print("Part2: {}\n", part2);
	return 0;
}

File: day11/day11.py
#!/usr/bin/env python3

class Map(object):
    NEIGHBORS = ((-1, 0), (-1, -1), (0, -1),
                 (1, -1),           (1,  0),
                 (1,  1), (0,   1), (-1, 1))

    def __init__(self, m):
        self.data = [[int(x) for x in row] for row in m.splitlines()]
        self.height = len(self.data)
        self.width = len(self.data[0])

    def inc(self, x, y):
        if self.data[y][x] < 0:
            return

        self.data[y][x] += 1
        if self.data[y][x] > 9:
            self.data[y][x] = -1
            for dx, dy in self.NEIGHBORS:
                nx = x + dx
                ny = y + dy
                if 0 <= nx < self.width and 0 <= ny < self.height:
                    self.inc(nx, ny)

    def step(self):
        # increase the level
        for y in range(self.height):
            for x in range(self.width):
                self.inc(x, y)

        # zero the flashed octopuses
        # NOTE: this is why they synchronize
        count = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.data[y][x] < 0:
                    count += 1
                    self.data[y][x] = 0

        return count

def solve(txt):
    m = Map(txt)
    i = 0
    part1 = 0
    while True:
        flashes = m.step()
        if i < 100:
            part1 += flashes
        if flashes == m.width * m.height:
            break
        i += 1
    return part1, i+1

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        part1, part2 = solve(txt)

        print("Part1:", part1)
        print("Part2:", part2)
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
